File: adminWeb/src/utils/slots.ts
/**
 * The windows a ticket can be served in — the browser's copy of the API's
 * `offered_slots`.
 *
 * It has to be expressed twice, for the same reason `_sla_order_case` mirrors
 * `sla_state` on the server: the list has to exist BEFORE the ticket does, so
 * there is no row to ask the API about. A vendor filling in the create form has
 * not created anything yet.
 *
 * The server is still the authority — `check_slot_bookable` re-derives this
 * list at commit time and refuses anything not on it. This copy exists so the
 * refusal is rare: a picker that only offers real windows beats a 400 arriving
 * after the form was filled in.
 *
 * Keep in step with `api/app/core/tickets.py` and `offered_slots` in
 * `api/app/features/tickets/service.py`. The four constants below are that
 * file's, value for value.
 */

import { formatTimeRange } from "@/utils/datetime";

/** Two-hour windows, 5 AM to 9 PM local. `SLOT_WINDOWS`. */
const SLOT_WINDOWS: ReadonlyArray<readonly [number, number]> = [
  [5, 7],
  [7, 9],
  [9, 11],
  [11, 13],
  [13, 15],
  [15, 17],
  [17, 19],
  [19, 21],
];

/** Nobody can be dispatched to an address in ten minutes. `SLOT_LEAD_MINUTES`. */
const SLOT_LEAD_MINUTES = 90;

/** IST, UTC+05:30. `SLOT_TIMEZONE_OFFSET_MINUTES`. */
const IST_OFFSET_MINUTES = 330;

/** How far ahead to walk. The server walks four days; so does this. */
const DAYS_AHEAD = 4;

export interface OfferedSlot {
  /** ISO instant — exactly what the API expects for `slotStart`. */
  start: string;
  end: string;
  /** `Thu 21 Aug` — the day heading. */
  day: string;
  /** `9:00–11:00 AM` IST, the same rendering `when_label` produces. */
  time: string;
}

/** Wall-clock IST parts for an instant, without pulling in a date library. */
function istParts(at: Date) {
  const shifted = new Date(at.getTime() + IST_OFFSET_MINUTES * 60_000);
  return {
    year: shifted.getUTCFullYear(),
    month: shifted.getUTCMonth(),
    date: shifted.getUTCDate(),
  };
}

/** The instant at which `hour` o'clock IST falls on that IST calendar day. */
function istInstant(
  parts: { year: number; month: number; date: number },
  hour: number,
  dayOffset: number
): Date {
  return new Date(
    Date.UTC(parts.year, parts.month, parts.date + dayOffset, hour) -
      IST_OFFSET_MINUTES * 60_000
  );
}

const DAY_FMT = new Intl.DateTimeFormat("en-IN", {
  weekday: "short",
  day: "2-digit",
  month: "short",
  timeZone: "Asia/Kolkata",
});

/** The windows are defined in IST, so they are labelled in IST — wherever the
 *  browser happens to be. Everything else in the console renders in local
 *  time, which for this market is the same clock. */
const IST_ZONE = "Asia/Kolkata";

/**
 * Every window this ticket could be served in, soonest first.
 *
 * Bounded at both ends, and both bounds matter: not sooner than the lead time,
 * and not later than the service level's deadline — a window past that is one
 * the company has already promised not to offer.
 *
 * An empty array is a real answer, not a bug: a 12-hour ticket raised at 22:00
 * has nothing left, and the form has to say so rather than render an empty menu.
 */
export function offeredSlots(
  serviceLevelHours: number,
  now: Date = new Date()
): OfferedSlot[] {
  const earliest = new Date(now.getTime() + SLOT_LEAD_MINUTES * 60_000);
  const latest = new Date(now.getTime() + serviceLevelHours * 3_600_000);
  const from = istParts(earliest);

  const out: OfferedSlot[] = [];
  for (let offset = 0; offset < DAYS_AHEAD; offset++) {
    for (const [fromHour, toHour] of SLOT_WINDOWS) {
      const begins = istInstant(from, fromHour, offset);
      if (begins < earliest || begins > latest) continue;
      const ends = istInstant(from, toHour, offset);
      out.push({
        start: begins.toISOString(),
        end: ends.toISOString(),
        // `Wed 19 Aug`, matching the server's `when_label`. Intl puts a comma
        // after the weekday; the server does not, and the two labels sit beside
        // each other on the ticket page.
        day: DAY_FMT.format(begins).replace(",", ""),
        time: formatTimeRange(begins, ends, IST_ZONE),
      });
    }
  }
  return out;
}

/** Today in IST, as `yyyy-mm-dd` — the `min` a date input needs. */
export function istToday(now: Date = new Date()): string {
  const { year, month, date } = istParts(now);
  return `${year}-${String(month + 1).padStart(2, "0")}-${String(date).padStart(2, "0")}`;
}

File: api/app/core/slots.py
"""Which two-hour windows a visit can still be booked into.

Two slices need this and hard rule 4 says they may not reach into each other:
`tickets` renders the customer's choices and validates what they pick, and
`jobs` has to ask the same question before letting a technician take a job that
has no time yet.

Its own module rather than a home in `core.tickets`, which holds the vocabulary
these are built from (`SLOT_WINDOWS`, `SLOT_LEAD_MINUTES`). `bookable_slots`
needs `core.coverage` for the cap statuses, and `core.coverage` already imports
`core.tickets` — putting this there would close the loop. Here the dependencies
run one way: slots → coverage → tickets.
"""

import datetime
import uuid

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.coverage import CAP_EXEMPT_STATUSES
from app.core.tickets import (
    SLOT_LEAD_MINUTES,
    SLOT_TIMEZONE_OFFSET_MINUTES,
    SLOT_WINDOWS,
)
from app.models.technician import TechnicianProfile
from app.models.ticket import Ticket

IST = datetime.timezone(datetime.timedelta(minutes=SLOT_TIMEZONE_OFFSET_MINUTES))


def _now() -> datetime.datetime:
    return datetime.datetime.now(datetime.timezone.utc)


def offered_slots(
    row: Ticket, *, now: datetime.datetime | None = None
) -> list[tuple[datetime.datetime, datetime.datetime]]:
    """Every window this ticket could still be served in, soonest first.

    Bounded at both ends, and both bounds matter:

      * not sooner than SLOT_LEAD_MINUTES from now — nobody can be dispatched to
        an address in ten minutes;
      * not later than `sla_due_at` — the service level says the slot must START
        within N hours of the ticket being raised, so a window past that is one
        the company has already promised not to offer.

    Because the list is generated from the window rather than filtered
    afterwards, a customer CANNOT pick a slot that breaches. That is the point:
    the constraint lives where the choice is made, not in a validator that has
    to say no to something already chosen.

    Empty is a real answer — a 12-hour ticket raised at 22:00 has nothing left
    to offer, and the page says so rather than showing an empty list.

    Pure and synchronous on purpose: `check_slot_bookable` runs it inside ticket
    creation, where there is no technician to consider and a database round trip
    would be one nobody needs. What a technician can actually serve is
    `bookable_slots` below.
    """
    now = now or _now()
    earliest = now + datetime.timedelta(minutes=SLOT_LEAD_MINUTES)
    latest = row.sla_due_at

    out: list[tuple[datetime.datetime, datetime.datetime]] = []
    # Walk local days, because the windows are local working hours. Three is
    # enough for the longest service level (48h) plus the day it spills into.
    start_day = earliest.astimezone(IST).date()
    for offset in range(4):
        day = start_day + datetime.timedelta(days=offset)
        for from_hour, to_hour in SLOT_WINDOWS:
            begins = datetime.datetime.combine(
                day, datetime.time(from_hour, tzinfo=IST)
            )
            ends = datetime.datetime.combine(day, datetime.time(to_hour, tzinfo=IST))
            if begins < earliest or begins > latest:
                continue
            out.append((begins, ends))
    return out


#: Sentinel for "use the row's own technician", so that `None` can mean the
#: honest thing — nobody is assigned — rather than being indistinguishable from
#: the argument being absent.
_UNSET: uuid.UUID = uuid.UUID(int=0)


async def bookable_slots(
    db: AsyncSession,
    row: Ticket,
    *,
    now: datetime.datetime | None = None,
    technician_id: uuid.UUID | None = _UNSET,
) -> list[tuple[datetime.datetime, datetime.datetime]]:
    """`offered_slots`, minus the ones the assigned technician cannot serve.

    A technician can now accept a job BEFORE a time exists, so by the moment the
    customer opens their link there may already be somebody committed to it. Ask
    that customer to pick from every window in the service level and they will
    cheerfully choose one their technician is already standing in somebody
    else's kitchen for.

    Two subtractions, and only when a technician is actually assigned:

      * **windows overlapping a job they already hold** — half-open, so a job
        ending at 14:00 does not block the window starting at 14:00;
      * **whole days their `daily_job_cap` is already spent on** — the same
        count `has_cap_room` uses, so a day the cap would refuse is never
        offered in the first place.

    Unassigned, this is exactly `offered_slots` and costs one branch. That is
    the common case, and the ordinary flow is untouched by any of it.

    `technician_id` overrides the one on the row, which is what `jobs.accept`
    needs: it asks the hypothetical — "if I took this, could I serve it?" —
    before any row has been written. A parameter beats assigning to `row` and
    putting it back, which would leave a dirty attribute on a tracked object for
    the next flush to find.

    ## Empty is a real answer, and now a worse one

    It already meant "the service level has run out". It can now also mean "the
    person who took this is full" — a dead end WE created and the customer
    cannot fix. The acceptance guard in `jobs.accept` makes it rare rather than
    impossible, because a technician can fill up after taking the job.
    """
    against = row.technician_id if technician_id is _UNSET else technician_id

    windows = offered_slots(row, now=now)
    if against is None or not windows:
        return windows

    held = (
        await db.execute(
            select(Ticket.slot_start, Ticket.slot_end).where(
                Ticket.company_id == row.company_id,
                Ticket.technician_id == against,
                # Not this ticket. Re-booking a visit must not be blocked by the
                # time it currently holds.
                Ticket.id != row.id,
                Ticket.deleted_at.is_(None),
                Ticket.status.not_in(CAP_EXEMPT_STATUSES),
                Ticket.slot_start.is_not(None),
            )
        )
    ).all()

    cap = await db.scalar(
        select(TechnicianProfile.daily_job_cap).where(
            TechnicianProfile.company_id == row.company_id,
            TechnicianProfile.id == against,
        )
    )

    spent: dict[datetime.date, int] = {}
    for start, _end in held:
        day = start.astimezone(IST).date()
        spent[day] = spent.get(day, 0) + 1

    def free(begins: datetime.datetime, ends: datetime.datetime) -> bool:
        if cap is not None and spent.get(begins.astimezone(IST).date(), 0) >= cap:
            return False
        return not any(begins < h_end and h_start < ends for h_start, h_end in held)

    return [w for w in windows if free(*w)]


# ── how a window is written down ─────────────────────────────────────────────
#
# Here rather than in the tickets slice because a job can now be accepted
# before a time exists, so `jobs` has to render one too — the acceptance
# notice names the slot, or says there is not one yet. Two slices formatting
# the same instant two ways is how a customer gets offered `14:00` and then
# told `2:00 PM`.


def clock(at: datetime.datetime) -> tuple[str, str]:
    """`('10:00', 'AM')` in IST — the two halves, so a range can share one.

    `.lstrip("0")` rather than `%-I`, which is a glibc extension and raises on
    Windows, where this very much does get run. `%I` never yields `"00"`, so
    stripping cannot empty the string.
    """
    local = at.astimezone(IST)
    return local.strftime("%I:%M").lstrip("0"), local.strftime("%p").upper()


def clock_range(start: datetime.datetime, end: datetime.datetime) -> str:
    """`10:00 AM–12:00 PM`, or `2:00–4:00 PM` when it stays inside one half.

    12-hour throughout, which is the house style taken from the approved
    prototypes — the technician app reads `4:00 PM` and its job data reads
    `2:00–4:00 PM`. A range that does not cross noon says the meridiem once.
    """
    start_hm, start_ap = clock(start)
    end_hm, end_ap = clock(end)
    if start_ap == end_ap:
        return f"{start_hm}–{end_hm} {end_ap}"
    return f"{start_hm} {start_ap}–{end_hm} {end_ap}"


def day_label(at: datetime.datetime) -> str:
    """`Thu 21 Aug` in IST — `when_label` without the clock.

    The day the WORK happens, in the day the technician experiences, which is
    the same reckoning the daily cap counts by. A UTC rendering would put a
    05:00 IST job on the previous evening and make the cap's arithmetic look
    wrong to whoever it refused.
    """
    return at.astimezone(IST).strftime("%a %d %b")


def when_label(start: datetime.datetime, end: datetime.datetime) -> str:
    """`Thu 21 Aug, 10:00 AM–12:00 PM`, in the customer's own timezone."""
    return f"{day_label(start)}, {clock_range(start, end)}"
